src/inspector_socket_server.cc

```diff
@@ -0,0 +1,471 @@
+
+    bool connected = false;
+    for (const auto& session : connected_sessions_) {
+      if (session.second->TargetId() == id) {
+        connected = true;
+        break;
+      }
+    }
+    if (!connected) {
+      std::string address = GetWsUrl(port_, id);
+      std::ostringstream frontend_url;
+      frontend_url << "chrome-devtools://devtools/bundled";
+      frontend_url << "/inspector.html?experiments=true&v8only=true&ws=";
+      frontend_url << address;
+      target_map["devtoolsFrontendUrl"] += frontend_url.str();
+      target_map["webSocketDebuggerUrl"] = "ws://" + address;
+    }
+  }
+  SendHttpResponse(socket, MapsToString(response));
+}
+
+bool InspectorSocketServer::Start(uv_loop_t* loop) {
+  loop_ = loop;
+  sockaddr_in addr;
+  uv_tcp_init(loop_, &server_);
+  uv_ip4_addr("0.0.0.0", port_, &addr);
+  int err = uv_tcp_bind(&server_,
+                        reinterpret_cast<const struct sockaddr*>(&addr), 0);
+  if (err == 0) {
+    err = uv_listen(reinterpret_cast<uv_stream_t*>(&server_), 1,
+                    SocketConnectedCallback);
+  }
+  if (err == 0 && connected_sessions_.empty()) {
+    PrintDebuggerReadyMessage(port_, delegate_->GetTargetIds());
+  }
+  if (err != 0 && connected_sessions_.empty()) {
+    fprintf(stderr, "Unable to open devtools socket: %s\n", uv_strerror(err));
+    uv_close(reinterpret_cast<uv_handle_t*>(&server_), nullptr);
+    return false;
+  }
+  return true;
+}
+
+void InspectorSocketServer::Stop(ServerCallback cb) {
+  if (closer_ == nullptr) {
+    closer_ = new Closer(this);
+  }
+  closer_->AddCallback(cb);
+
+  uv_handle_t* handle = reinterpret_cast<uv_handle_t*>(&server_);
+  if (uv_is_active(handle)) {
+    closer_->IncreaseExpectedCount();
+    uv_close(reinterpret_cast<uv_handle_t*>(&server_), ServerClosedCallback);
+  }
+  closer_->NotifyIfDone();
+}
+
+void InspectorSocketServer::TerminateConnections(ServerCallback cb) {
+  if (closer_ == nullptr) {
+    closer_ = new Closer(this);
+  }
+  closer_->AddCallback(cb);
+  std::map<int, SocketSession*> sessions;
+  std::swap(sessions, connected_sessions_);
+  for (const auto& session : sessions) {
+    int id = session.second->Id();
+    session.second->Close(true, closer_);
+    delegate_->EndSession(id);
+  }
+  closer_->NotifyIfDone();
+}
+
+bool InspectorSocketServer::TargetExists(const std::string& id) {
+  const std::vector<std::string>& target_ids = delegate_->GetTargetIds();
+  const auto& found = std::find(target_ids.begin(), target_ids.end(), id);
+  return found != target_ids.end();
+}
+
+void InspectorSocketServer::Send(int session_id, const std::string& message) {
+  auto session_iterator = connected_sessions_.find(session_id);
+  if (session_iterator != connected_sessions_.end()) {
+    session_iterator->second->Send(message);
+  }
+}
+
+// static
+void InspectorSocketServer::ServerClosedCallback(uv_handle_t* server) {
+  InspectorSocketServer* sserver = InspectorSocketServer::From(server);
```

Comment:
Suggestion: s/sserver/socket_server/ - the difference between server and sserver is very subtle.